src/shared/cryptography.c

```diff
@@ -0,0 +1,536 @@
+                                    error_message_size,
+                                    "CryptCATAdminCalcHashFromFileHandle failed trying to calculate hash with error "
+                                    "%lu for '%S': %s",
+                                    result,
+                                    full_path,
+                                    win_strerror(result));
+                        os_free(*hash);
+                    }
+                }
+            } else {
+                result = GetLastError();
+                os_snprintf(error_message,
+                            error_message_size,
+                            "CryptCATAdminCalcHashFromFileHandle failed trying to get the hash size with error %lu for "
+                            "'%S': %s",
+                            result,
+                            full_path,
+                            win_strerror(result));
+            }
+        } else {
+            result = ERROR_FILE_NOT_FOUND;
+            last_error = GetLastError();
+            os_snprintf(error_message,
+                        error_message_size,
+                        "CreateFileW failed with error %lu for '%S': %s",
+                        last_error,
+                        full_path,
+                        win_strerror(last_error));
+        }
+        // Close file handle.
+        CloseHandle(handle_file);
+
+    } else {
+        result = ERROR_INVALID_DATA;
+        last_error = GetLastError();
+        os_snprintf(error_message,
+                    error_message_size,
+                    "GetFullPathNameW failed with error %lu for '%S'. %s",
+                    last_error,
+                    path,
+                    win_strerror(last_error));
+    }
+
+    return result;
+}
+
+
+
+DWORD check_ca_available(const char* ca) {
+    // Check if the CA is available in the system.
+    // If the CA is not available, the function returns some error code.
+    // If the CA is available, the function returns ERROR_SUCCESS.
+    DWORD result = ERROR_INVALID_DATA;
+    HCERTSTORE cert_store = NULL;
+    PCCERT_CONTEXT cert_context = NULL;
+    char ca_name[OS_MAXSTR];
+
+    // Check if the CA name is not NULL.
+    if (NULL != ca) {
+        const size_t ca_length = strlen(ca);
+        // Check if the CA name is not too long.
+        if (ca_length < OS_MAXSTR) {
+            // Open the certificate store.
+            cert_store = CertOpenSystemStore(0, "ROOT");
+            // Check if the certificate store was opened successfully.
+            if (cert_store) {
+                // Get the first certificate in the store.
+                cert_context = CertEnumCertificatesInStore(cert_store, NULL);
+
+                while (cert_context) {
+                    // Get the certificate's CN friendly name.
+                    if (CertGetNameString(cert_context,
+                                        CERT_NAME_FRIENDLY_DISPLAY_TYPE,
```

Comment:
Maybe we should use `CERT_NAME_SIMPLE_DISPLAY_TYPE` instead.

I've compared the output of the two flags in a Win 8.1 VM.

<details>

<summary>CERT_NAME_FRIENDLY_DISPLAY_TYPE</summary>

```
CA name: Microsoft Root Certificate Authority
CA name: Thawte Timestamping CA
CA name: Microsoft Root Authority
CA name: Microsoft Root Certificate Authority 2011
CA name: Microsoft Authenticode(tm) Root
CA name: Microsoft Root Certificate Authority 2010
CA name: Microsoft Timestamp Root
CA name: VeriSign Time Stamping CA
CA name: Sectigo (UTN Object)
CA name: DigiCert Global Root G2
CA name: DigiCert Trusted Root G4
CA name: DST Root CA X3
CA name: GlobalSign Root CA - R3
CA name: DigiCert Baltimore Root
CA name: Sectigo (AAA)
CA name: ISRG Root X1
CA name: GlobalSign Root CA - R1
CA name: Sectigo (formerly Comodo CA)
CA name: Starfield Class 2 Certification Authority
CA name: DigiCert
CA name: TWCA Global Root CA
CA name: Entrust.net
CA name: VeriSign Class 3 Public Primary CA
CA name: DigiCert
CA name: VeriSign
CA name: Go Daddy Root Certificate Authority – G2
CA name: Sectigo
CA name: Go Daddy Class 2 Certification Authority
CA name: QuoVadis Root CA 2 G3
CA name: DigiCert
CA name: Sectigo (AddTrust)
```
</details>

<details>

<summary>CERT_NAME_SIMPLE_DISPLAY_TYPE</summary>

```
CA name: Microsoft Root Certificate Authority
CA name: Thawte Timestamping CA
CA name: Microsoft Root Authority
CA name: Microsoft Root Certificate Authority 2011
CA name: Microsoft Authenticode(tm) Root Authority
CA name: Microsoft Root Certificate Authority 2010
CA name: Copyright (c) 1997 Microsoft Corp.
CA name: NO LIABILITY ACCEPTED, (c)97 VeriSign, Inc.
CA name: UTN-USERFirst-Object
CA name: DigiCert Global Root G2
CA name: DigiCert Trusted Root G4
CA name: DST Root CA X3
CA name: GlobalSign
CA name: Baltimore CyberTrust Root
CA name: AAA Certificate Services
CA name: ISRG Root X1
CA name: GlobalSign Root CA
CA name: COMODO RSA Certification Authority
CA name: Starfield Class 2 Certification Authority
CA name: DigiCert Global Root CA
CA name: TWCA Global Root CA
CA name: Entrust Root Certification Authority - G2
CA name: Class 3 Public Primary Certification Authority
CA name: DigiCert High Assurance EV Root CA
CA name: VeriSign Class 3 Public Primary Certification Authority - G5
CA name: Go Daddy Root Certificate Authority - G2
CA name: USERTrust RSA Certification Authority
CA name: Go Daddy Class 2 Certification Authority
CA name: QuoVadis Root CA 2 G3
CA name: DigiCert Assured ID Root CA
CA name: AddTrust External CA Root

```
</details>

We can see that some `DigiCert` entries are expanded and if we don't use the full name, we might think that the right CA is installed when it isn't.